src/Common/src/System/IO/StreamHelpers.cs

```diff
@@ -0,0 +1,64 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.IO
+{
+    /// <summary>Provides methods to help in the implementation of Stream-derived types.</summary>
+    internal static class StreamHelpers
+    {
+        /// <summary>
+        /// Provides an implementation usable as an override of Stream.CopyToAsync but that uses the shared
+        /// ArrayPool for the intermediate buffer rather than allocating a new buffer each time.
+        /// </summary>
+        /// <remarks>
+        /// If/when the base CopyToAsync implementation is changed to use a pooled buffer, 
+        /// this will no longer be necessary.
+        /// </remarks>
+        public static Task ArrayPoolCopyToAsync(Stream source, Stream destination, int bufferSize, CancellationToken cancellationToken)
+        {
+            Debug.Assert(source != null);
+
+            if (destination == null)
+            {
+                throw new ArgumentNullException("destination");
+            }
+            if (bufferSize <= 0)
+            {
+                throw new ArgumentOutOfRangeException("bufferSize", bufferSize, SR.ArgumentOutOfRange_NeedPosNum);
+            }
+            if (!source.CanRead)
+            {
+                throw new NotSupportedException(SR.NotSupported_UnreadableStream);
+            }
+            if (!destination.CanWrite)
+            {
+                throw new NotSupportedException(SR.NotSupported_UnwritableStream);
+            }
+
+            byte[] buffer = ArrayPool<byte>.Shared.Rent(bufferSize); // rented here to reduce size of async state machine boxed object
+            return ArrayPoolCopyToAsyncInternal(source, destination, buffer, cancellationToken);
+        }
+
+        private static async Task ArrayPoolCopyToAsyncInternal(Stream source, Stream destination, byte[] buffer, CancellationToken cancellationToken)
+        {
+            try
+            {
+                int bytesRead;
+                while ((bytesRead = await source.ReadAsync(buffer, 0, buffer.Length, cancellationToken).ConfigureAwait(false)) != 0)
+                {
+                    await destination.WriteAsync(buffer, 0, bytesRead, cancellationToken).ConfigureAwait(false);
+                }
```

Comment:
I considered that but decided against it for the same reasons we didn't do so in the original Stream.CopyToAsync implementation.  It leads to interesting concurrency issues that we decided to avoid in these core methods.  For example, if the two concurrent read and write operations throw, which exception should we propagate, and what does that mean if there were important information in the other?  And if the operations would actually end up consuming a thread pool thread to do a synchronous operation, do we really want to potentially be blocking two thread pool threads rather than one?  Etc.  If someone determines that they would really benefit from such overlapping, they can make that decision at the app level (where they have the ability to measure impact at the app-level) and write the few lines like you have themselves.  Thanks, though.